File: ModuloZonas/src/main/java/ec/edu/espe/zonas/infrastructure/security/PermissionsCacheService.java
package ec.edu.espe.zonas.infrastructure.security;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.amqp.core.ExchangeTypes;
import org.springframework.amqp.core.Message;
import org.springframework.amqp.rabbit.annotation.Exchange;
import org.springframework.amqp.rabbit.annotation.Queue;
import org.springframework.amqp.rabbit.annotation.QueueBinding;
import org.springframework.amqp.rabbit.annotation.RabbitListener;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Autorización PULL para el microservicio de zonas.
 *  - Pide a gestion-usuarios (endpoint interno) los permisos de un rol filtrados
 *    por este servicio, y los cachea en memoria con TTL.
 *  - Escucha eventos 'role_permissions.changed' por RabbitMQ para invalidar la caché.
 */
@Service
public class PermissionsCacheService {

    private static final Logger log = LoggerFactory.getLogger(PermissionsCacheService.class);
    private static final long TTL_MS = 5 * 60_000L;
    private static final String SERVICE_ID = "zonas-service";

    private record CacheEntry(List<String> perms, long exp) {}

    private final ConcurrentHashMap<String, CacheEntry> cache = new ConcurrentHashMap<>();
    private final RestClient restClient;
    private final ObjectMapper objectMapper;
    private final String internalUrl;
    private final String internalKey;

    public PermissionsCacheService(
            ObjectMapper objectMapper,
            @Value("${app.authz.internal-url:http://gestion-usuarios:3001/api/usuarios}") String internalUrl,
            @Value("${app.authz.internal-key:}") String internalKey) {
        this.objectMapper = objectMapper;
        this.internalUrl = internalUrl.replaceAll("/+$", "");
        this.internalKey = internalKey;
        this.restClient = RestClient.create();
    }

    /** Devuelve los permisos del rol para este servicio (cacheados). [] ante fallo. */
    @SuppressWarnings("unchecked")
    public List<String> getPermissions(String role) {
        CacheEntry entry = cache.get(role);
        if (entry != null && entry.exp() > System.currentTimeMillis()) {
            return entry.perms();
        }

        try {
            Map<String, Object> body = restClient.post()
                    .uri(internalUrl + "/internal/role-permissions/resolve")
                    .header("x-internal-key", internalKey)
                    .header("Content-Type", "application/json")
                    .body(Map.of("role", role, "serviceId", SERVICE_ID))
                    .retrieve()
                    .body(Map.class);

            List<String> perms = body != null && body.get("permissions") != null
                    ? (List<String>) body.get("permissions")
                    : List.of();
            cache.put(role, new CacheEntry(perms, System.currentTimeMillis() + TTL_MS));
            return perms;
        } catch (Exception e) {
            log.error("No se pudieron resolver permisos de '{}': {}", role, e.getMessage());
            return List.of(); // deny-by-default
        }
    }

    public void invalidate(String role) {
        cache.remove(role);
        log.info("Caché de permisos invalidada para rol '{}'", role);
    }

    /**
     * Consume eventos de invalidación. Cola anónima (una por instancia) enlazada al
     * exchange topic de authz con la routing key de cambios de permisos.
     */
    @RabbitListener(bindings = @QueueBinding(
            value = @Queue(autoDelete = "true", durable = "false"),
            exchange = @Exchange(value = "${app.authz.exchange:authz_exchange}",
                    type = ExchangeTypes.TOPIC, durable = "true"),
            key = "role_permissions.changed"))
    public void onAuthzEvent(Message message) {
        // Recibimos el Message crudo para NO pasar por el Jackson2MessageConverter global
        // (el publisher NestJS envía bytes JSON sin content_type).
        try {
            Map<String, Object> evt = objectMapper.readValue(message.getBody(), Map.class);
            Object service = evt.get("service");
            if (service == null || SERVICE_ID.equals(service)) {
                invalidate((String) evt.get("role"));
            }
        } catch (Exception e) {
            log.warn("Evento authz no procesable: {}", e.getMessage());
        }
    }
}

File: assignment-service/src/auth/permissions-cache.service.ts
import {
  Injectable,
  Logger,
  OnModuleInit,
  OnModuleDestroy,
} from '@nestjs/common';
import { HttpService } from '@nestjs/axios';
import { ConfigService } from '@nestjs/config';
import { firstValueFrom } from 'rxjs';
import * as amqp from 'amqplib';

interface CacheEntry {
  perms: string[];
  exp: number;
}

/**
 * Autorización PULL para este microservicio:
 *  - Al necesitar los permisos de un rol, los pide a gestion-usuarios (endpoint interno)
 *    filtrados por este servicio, y los cachea en memoria con TTL.
 *  - Escucha eventos 'role_permissions.changed' por RabbitMQ para invalidar la caché
 *    cuando cambian los permisos de un rol (la caché con TTL es la red de seguridad).
 */
@Injectable()
export class PermissionsCacheService implements OnModuleInit, OnModuleDestroy {
  private readonly logger = new Logger(PermissionsCacheService.name);

  /** Id de este servicio; debe coincidir con la columna `service` de los permisos. */
  private readonly SERVICE_ID = 'assignment-service';
  private readonly TTL_MS = 5 * 60_000;

  private cache = new Map<string, CacheEntry>(); // key = nombre de rol
  private readonly internalUrl: string;
  private readonly internalKey: string;

  // RabbitMQ (consumidor de invalidación)
  private connection: any = null;
  private channel: any = null;
  private reconnectTimeout: NodeJS.Timeout | null = null;
  private readonly exchange: string;

  constructor(
    private readonly http: HttpService,
    private readonly config: ConfigService,
  ) {
    this.internalUrl =
      this.config.get<string>('USUARIOS_INTERNAL_URL') ??
      'http://gestion-usuarios:3001/api/usuarios';
    this.internalKey = this.config.get<string>('INTERNAL_API_KEY') ?? '';
    this.exchange = this.config.get<string>('AUTHZ_EXCHANGE') ?? 'authz_exchange';
  }

  async onModuleInit() {
    await this.connectRabbit();
  }

  /**
   * Devuelve los permisos del rol para este servicio (cacheados).
   * En caso de fallo del pull, devuelve [] (deny-by-default).
   */
  async getPermissions(role: string): Promise<string[]> {
    const hit = this.cache.get(role);
    if (hit && hit.exp > Date.now()) {
      return hit.perms;
    }

    try {
      const { data } = await firstValueFrom(
        this.http.post(
          `${this.internalUrl}/internal/role-permissions/resolve`,
          { role, serviceId: this.SERVICE_ID },
          { headers: { 'x-internal-key': this.internalKey }, timeout: 3000 },
        ),
      );
      const perms: string[] = data?.permissions ?? [];
      this.cache.set(role, { perms, exp: Date.now() + this.TTL_MS });
      return perms;
    } catch (error: any) {
      this.logger.error(
        `No se pudieron resolver permisos de '${role}': ${error?.message}`,
      );
      return []; // deny-by-default
    }
  }

  invalidate(role: string) {
    this.cache.delete(role);
    this.logger.log(`Caché de permisos invalidada para rol '${role}'`);
  }

  invalidateAll() {
    this.cache.clear();
  }

  // ───────────────────────── RabbitMQ ─────────────────────────

  private async connectRabbit(): Promise<void> {
    const host = this.config.get('RABBITMQ_HOST');
    const port = this.config.get('RABBITMQ_PORT');
    const user = this.config.get('RABBITMQ_USER');
    const pass = this.config.get('RABBITMQ_PASSWORD');
    const url = `amqp://${user}:${pass}@${host}:${port}`;

    try {
      this.connection = await amqp.connect(url);
      this.channel = await this.connection.createChannel();
      await this.channel.assertExchange(this.exchange, 'topic', { durable: true });

      // Cola exclusiva por instancia: cada réplica invalida su propia caché.
      const q = await this.channel.assertQueue('', { exclusive: true });
      await this.channel.bindQueue(q.queue, this.exchange, 'role_permissions.changed');

      await this.channel.consume(
        q.queue,
        (msg: any) => {
          if (!msg) return;
          try {
            const evt = JSON.parse(msg.content.toString());
            // Invalida si el evento afecta a este servicio (o a todos: service null)
            if (!evt.service || evt.service === this.SERVICE_ID) {
              this.invalidate(evt.role);
            }
          } catch (e: any) {
            this.logger.warn(`Evento authz no procesable: ${e?.message}`);
          }
        },
        { noAck: true },
      );

      this.logger.log('✅ Suscrito a eventos authz para invalidación de caché');

      this.connection.on('close', () => {
        this.channel = null;
        this.connection = null;
        this.scheduleReconnect();
      });
      this.connection.on('error', (err: any) => {
        this.logger.error(`❌ Error RabbitMQ (authz cache): ${err.message}`);
      });
    } catch (error: any) {
      this.logger.error(
        `❌ No se pudo suscribir a RabbitMQ (authz cache): ${error?.message}`,
      );
      this.scheduleReconnect();
    }
  }

  private scheduleReconnect() {
    if (this.reconnectTimeout) clearTimeout(this.reconnectTimeout);
    this.reconnectTimeout = setTimeout(() => this.connectRabbit(), 5000);
  }

  async onModuleDestroy() {
    if (this.reconnectTimeout) clearTimeout(this.reconnectTimeout);
    try {
      if (this.channel) await this.channel.close();
      if (this.connection) await this.connection.close();
    } catch {
      // Ignoramos errores al cerrar
    }
  }
}
